src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java

```diff
@@ -106,28 +189,127 @@ protected void appendFieldEqualityClause(JoinDefinition joinDefinition) {
       }
 
       prev = curr;
     }
   }
 
+  private void appendOnExpressionClause() {
+    JoinCondition.OnExpression onExpression = (JoinCondition.OnExpression) joinDefinition.getCondition();
+
+    JoinStage left = joinDefinition.getStages().get(0);
+    JoinStage right = joinDefinition.getStages().get(1);
+
+    // Append Join Statement for these 2 stages
+    // ...<left_table> <join_type> JOIN <left_table> ON ...
+    appendFullTableNameAndAlias(left.getStageName());
+    builder.append(SPACE);
+    appendJoinStatement(left, right);
+
+    String leftAlias = stageToTableAliasMap.get(left.getStageName());
+    String rightAlias = stageToTableAliasMap.get(right.getStageName());
+
+    String leftStageIdentifier = onExpression.getDatasetAliases()
+      .getOrDefault(left.getStageName(), left.getStageName());
+    String rightStageIdentifier = onExpression.getDatasetAliases()
+      .getOrDefault(right.getStageName(), right.getStageName());
+
+    appendOnExpressionClause(onExpression.getExpression(),
+                             leftStageIdentifier, leftAlias,
+                             rightStageIdentifier, rightAlias);
+  }
+
+  /**
+   * Adds expression clause for these 2 stages.
+   * <p>
+   * Executes replace operation to replace table names in the supplied string for BigQuery-safe table aliases.
+   *
+   * @param expression       expression to manipulate
+   * @param leftIdentifier   Left Stage name
+   * @param leftReplacement  Left Stage replacement alias
+   * @param rightIdentifier  Right Stage Name
+   * @param rightReplacement Right Stage replacement alias
+   */
+  @VisibleForTesting
+  protected void appendOnExpressionClause(String expression,
+                                          String leftIdentifier, String leftReplacement,
+                                          String rightIdentifier, String rightReplacement) {
+
+    // Find a temporary string to use as a placeholder. This is used for example, if you have 2 stages:
+    // Stage "A" -> Alias "B"
+    // Stage "B" -> Alias "A"
+    // Since we do the replacement one-at-a-time, we need this intermediate replacement to ensure the flip does not
+    // corrupt the original query.
+    String temp = UUID.randomUUID().toString();
+    while (temp.equals(leftIdentifier) || temp.equals(rightIdentifier)) {
+      temp = UUID.randomUUID().toString();
+    }
+
+    String replacement1 = String.format(TABLE_NAME_PATTERN, Pattern.quote(leftIdentifier));
+    String replacement2 = String.format(TABLE_NAME_PATTERN, Pattern.quote(rightIdentifier));
+    String replacement3 = String.format(TABLE_NAME_PATTERN, Pattern.quote(temp));
+
+
+    // Replace the left idenfier with the left alias, and the right identifier with the right alias.
+    // Note that Left is transformer to the temporary placeholder before flipping to the right value.
+    String replaced = expression
+      .replaceAll(replacement1, "$1" + Matcher.quoteReplacement(temp) + "$5$6")
```

Comment:
nit: consider using named groups, it would be a bit more readable. Also you can join groups 5 and 6.